src/charm.py

```diff
@@ -606,37 +633,41 @@ def _on_nrpe_targets_changed(self, event: Optional[NrpeTargetsChangedEvent]):
             )
 
     def _set_status(self):
-        messages = []
-        if (
-            (self._stored.have_grafana or self._stored.have_gagent)
-            and not self._stored.have_dashboards
-        ) or (  # pyright: ignore
-            self._stored.have_dashboards
-            and not (self._stored.have_grafana or self._stored.have_gagent)  # pyright: ignore
-        ):
-            messages.append("one of (Grafana|dashboard|grafana-agent)")
+        # Put charm in blocked status if all incoming relations are missing
+        # We could obtain the set of active relations with:
+        # {k for k, v in self.model.relations.items() if v}
+        # but that would incur a relation-list and multiple relation-get calls, so building up from
+        # stored state instead.
+        active_relations = {
+            rel
+            for (rel, indicator) in [
+                ("downstream-grafana-dashboard", self._stored.have_grafana),
+                ("cos-agent", self._stored.have_gagent),
+                ("dashboards", self._stored.have_dashboards),
+                ("downstream-logging", self._stored.have_loki),
+                ("filebeat", self._stored.have_filebeat),
+                ("monitors", self._stored.have_nrpe),
+                ("downstream-prometheus-scrape", self._stored.have_prometheus),
+                ("prometheus-target", self._stored.have_targets),
+            ]
+            if indicator
+        }
+
+        if not set(self.mandatory_relation_pairs.keys()).intersection(active_relations):
+            self.unit.status = BlockedStatus(
+                "Missing incoming relations: {}".format(
+                    "|".join(self.mandatory_relation_pairs.keys())
```

Comment:
```suggestion
                    "|".join(self.mandatory_relation_pairs.keys())
```

we should be only listing the difference between what should be and actually is, not the full list, to avoid cluttering.